### PR TITLE
Add achievements relation to CourseUser

### PR DESCRIPTION
**Pull request by minhtule**
Needed for #496

### CODE REVIEW

**Comment by lowjoel** [comment]
Build failed, and add a spec for it.

**Comment by minhtule** [comment]
@lowjoel Sorry, I forgot this is a many-to-many relation. Please review again!

File: app/models/course/achievement.rb [hunk 1]
```diff
@@ -4,6 +4,9 @@ class Course::Achievement < ActiveRecord::Base
   after_initialize :set_defaults, if: :new_record?
 
   belongs_to :course, inverse_of: :achievements
+  has_many :user_achievements, class_name: Course::UserAchievement.name,
```
Comment (lowjoel) [inline app/models/course/achievement.rb]: course_user_achievements @allenwq what do you think?

File: app/models/course_user.rb [hunk 2]
```diff
@@ -34,6 +34,9 @@ def experience_points
       sum(:points_awarded)
     end
   end
+  has_many :user_achievements, class_name: Course::UserAchievement.name, inverse_of: :course_user,
```
Comment (lowjoel) [inline app/models/course_user.rb]: name this the same as in Course::Achievement, when we decide on a name.

**Comment by minhtule** [comment]
@joel please review again!

File: db/migrate/20151011151130_create_course_user_achievements.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class CreateCourseUserAchievements < ActiveRecord::Migration
+  def change
+    create_table :course_user_achievements do |t|
+      t.references :course_user
+      t.references :achievement, foreign_key: { references: :course_achievements }
+      t.datetime :obtained_at, null: false
```
Comment (lowjoel) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: Isn't this just the created_at column?

Comment (minhtule) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: There are 2 timestamps: one is when the student submits his/her mission and another is when the achievement is actually awarded i.e. the mission is graded by a tutor. Do I understand correctly @allenwq ?

Comment (lowjoel) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: If there is a link between user and achievement, the achievement is automatically awarded. that's what is meant by join table right?

Comment (allenwq) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: @lowjoel This field has different meaning with `created_at`, `created_at` is the time when user really get the achievement, but the `obtained_at` should be set to the time when student submit his assessment.  Thus TA's performance won't affect student's leaderboard ranking. 

Maybe I should give you an example:

S submits his mission on Monday
T grades S's submission on Tuesday and S get the achievement on Tuesday
S's achievement's `obtained_at` should be set to Monday

Comment (lowjoel) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: I don't think we use the time that the user actually got the achievement right?

How do we sort the leaderboard?

Comment (allenwq) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: We sort the leaderboard by their achievements counts first, then obtained_at. obtained_at is the date when student submits his assessment.

Comment (lowjoel) [inline db/migrate/20151011151130_create_course_user_achievements.rb]: OK.

**Pull request merged**